---
title: "MSGenova (Clean version) Rmakrdown Report"
author: "Gabri√®le Adam"
date: "2022-08-16"
output:
    html_document:
        highlight: tango
        number_sections: no
        theme: default
        toc: yes
        toc_depth: 3
        toc_float:
            collapsed: no
            smooth_scroll: yes
params:
  DP.min: DP.min
  AR.min: AR.min
  result_dir: result_dir
  rmd: "report.Rmd"
---

```{r setup, include=FALSE}
workspace<-params$result_dir
knitr::opts_chunk$set(root.dir=workspace, echo = FALSE , warning=FALSE, error=FALSE, message=FALSE)
```

```{r load_library, echo=FALSE, warning=FALSE, error=FALSE}
library(tidyverse)
library(DT)
library(ggplot2)
library(plyr)
library(dplyr)
library(yaml)
vector_calling<-params$vector

DP.min<-5
AR.min<-0.2

sheet<-params$sample
trimmo<- params$trimmo
flagstat<- params$flagstat
haplotype_caller<- params$haplotype_caller

config_advanced_file<-paste0(workspace,"/00_logs/config_advanced")
config_ad<-read_yaml(config_advanced_file)
```


## Trimming parameters

```{r trimmomatic}
trimmo<-paste0(workspace,"/01_sequence_qc/log/trimmomatic.log")
dat<-read.table(text=gsub(" ","\t",readLines(trimmo)),  header=TRUE)
DT::datatable(dat)
```

## Mapping parameters

### Flagstat results

```{r flagstat}
library(tidyverse)
file <- paste0(workspace,"/02_mapping/flagstat/concatenate_flagstat.txt")
if(file.exists(file)){
    flagstat <- read.table(file, sep="\t", header=FALSE) %>% 
      dplyr::mutate(mapped=V3-V4, unmapped=V2-V3, supplementary=V4, 
                    mapped.ratio=round(mapped/(mapped+unmapped)*100,2), 
                    unmapped.ratio=round(unmapped/(mapped+unmapped)*100,2),
                    supplementary.ratio=round(supplementary/(mapped+unmapped)*100,2)) %>% dplyr::select(-V2, -V3, -V4)
    names(flagstat) <- c("sample", "mapped", "unmapped", "supplementary", "mapped.ratio", "unmapped.ratio", "supplementary.ratio")

  } 

DT::datatable(flagstat, options = list(rownames = TRUE, pageLength = 10, scrollX = T, dom = 'tip'),class = 'cell-border stripe')
#count
nbr.reads.melt <- flagstat %>% dplyr::select(sample, mapped, unmapped, supplementary) %>% reshape2::melt(id = "sample", value.name = "nbr.reads")
summary.nbr <- nbr.reads.melt %>% group_by(variable) %>% summarise(median=median(nbr.reads), max=max(nbr.reads), min=min(nbr.reads))
nbr.reads.melt <- left_join(nbr.reads.melt, summary.nbr, by = "variable") %>% mutate(x.label=paste0(variable, "\nmedian=", median, "\nmax=", max, "\nmin=", min))
ggplot(data = nbr.reads.melt, aes(y=nbr.reads, x=x.label)) + geom_boxplot()  + xlab("") + ylab("number of reads") + ggtitle("Boxplot representing the number of mapped/supplementary/unmapped reads")+theme_bw()
#ratio
ratio.reads.melt <- flagstat %>% dplyr::select(sample, mapped.ratio, supplementary.ratio) 
names(ratio.reads.melt) <- c("sample", "mapped", "supplementary")
ratio.reads.melt <- ratio.reads.melt %>% reshape2::melt(id = "sample", value.name = "ratio")
summary.ratio <- ratio.reads.melt %>% group_by(variable) %>% summarise( median=round(median(ratio),2), max=round(max(ratio),2), min=round(min(ratio),2))
ratio.reads.melt <- left_join(ratio.reads.melt, summary.ratio, by = "variable") %>% mutate(x.label=paste0(variable, "\nmedian=", median, "%\nmax=", max, "%\nmin=", min, "%"))
ggplot(data = ratio.reads.melt) + geom_boxplot(aes(y=ratio, x=x.label)) + xlab("") + ylab("% reads") + scale_y_continuous(breaks = seq(0, max(ratio.reads.melt$ratio)+10, 20)) + ggtitle("Boxplot representing the percentage of mapped reads and supplementary alignments")+theme_bw()

```

### Coverage analysis
```{r coverage}
cov.list <- lapply(sample_sheet$sample, function(x){
  file <- paste0(workspace,"/02_mapping/coverage/", x, ".coverage")
  if(file.exists(file)){
    
    read.table(file, sep="\t") %>%  dplyr::filter(V4 > 1)  %>% dplyr::mutate(sample = x)
  }
})

cov <- purrr::reduce(cov.list, rbind) %>% dplyr::mutate(nbr_reads=dplyr::if_else(V4 >= 15, ">15X", "<15X"))
colnames(cov) <- c("rname",	"startpos",	"endpos",	"numreads",	"covbases",	"coverage",	"meandepth",	"meanbaseq",	"meanmapq", "sample", "nbr_reads")
#cov <- cov %>% left_join(bed.tab, by=c("rname","startpos"="start","endpos"="stop")) 
cov.summary <- cov %>% group_by(rname) %>% summarise(min=min(numreads), max=max(numreads), median=round(median(numreads)))
cov <- left_join(cov, cov.summary, by = c("rname") )%>% mutate(x.label=paste0(rname, "\nmedian=", median, "\nmax=", max, "\nmin=", min)) 
for (i in unique(sample_sheet$sample)){
cat("####",i,"\n")
g<-ggplot(filter(cov,sample==i)) + geom_boxplot(aes(y=numreads, x=x.label)) + xlab("") + ylab("#reads") + ggtitle("Plot coverage")+facet_wrap(. ~ sample,scales = "free")+theme(axis.text.x=element_text(angle=90,hjust=1))+theme_bw()
print(g)
cat('\n\n')
}
```

### Depth analysis
```{r depth_analysis}
cov.list <- lapply(sample_sheet$sample, function(x){

  file <- paste0(workspace,"/02_mapping/depth/", x, ".depth")
  if(file.exists(file)){
    read.table(file, sep="\t")  %>% dplyr::mutate(plant = x)
  }
})

depth <- purrr::reduce(cov.list, rbind)
colnames(depth) <- c("rname", "position", "depth", "sample")
#depth <- depth %>% left_join(bed.tab, by="rname")  %>% filter(position >=start, position <= stop)
heading<-as.character(unique(sample_sheet$sample))
for (i in unique(cov$rname)){
 cat("####",i,"\n")
  p <- ggplot(data = filter(depth,rname== i)) + geom_line(aes(x=position, y=depth, color=sample))   + theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylab("Depth of coverage")+facet_grid(~rname, scales = "free") 
  print(p)
cat('\n\n')
}

```

```{r proportion_of, results='asis',echo=FALSE, eval=FALSE, warning=FALSE, error=FALSE,results=FALSE}
for (i in unique(bed.tab$target)){
 cat("####",i,"\n")
tmp <- cov  %>% filter(target== i) %>% group_by(nbr_reads, target, amplicon) %>% count(name = "#well") 
g1<-ggplot(data = tmp) + geom_bar(aes(x=amplicon, y=`#well`, fill=nbr_reads), stat = "identity") + xlab("") +facet_grid(~target, scales = "free")
g2<-cov %>% filter(target== i) %>% group_by(well, nbr_reads,target) %>% count(name = "amplicon") %>% filter(nbr_reads == ">15X") %>% group_by(amplicon,target) %>% count(name = "#well") %>% ggplot(aes(x = amplicon, y = `#well`)) + geom_bar(stat = "identity") + xlab("nbr of amplicon covered at >15X")
print(g1)
print(g2)
cat('\n\n')
}
```

## Study of SNV and Indels calling

```{r define_r_fct_multiallele, echo=FALSE}
vcfR_split_multiAlt <- function(vcfR.tid, nbrAlt=NULL){

  vcfR.tid.tmp <- 
    dplyr::mutate(vcfR.tid, nb.allele=unlist(lapply(str_split(string = vcfR.tid$gt_AD, pattern = ","), function(x){return(length(x))}))) %>%
    dplyr::mutate(nb.allele=nb.allele-1)
  
  vcfR.tid.list <- lapply(unique(vcfR.tid.tmp$nb.allele), function(x){
    
    tmp <- vcfR.tid.tmp %>% dplyr::filter(nb.allele == x) %>%
      tidyr::separate(col = gt_AD, into = c("gt_ref_AD", paste0("gt_alt_AD",1:x)), sep = ",", remove = FALSE) %>%
      tidyr::separate(col = ALT,   into = paste0("ALT" ,1:x), sep = ",", remove = FALSE) %>%
      tidyr::separate(col = AC,    into = paste0("AC" ,1:x),  sep = ",", remove = FALSE) %>%
      tidyr::separate(col = AF,    into = paste0("AF" ,1:x),  sep = ",", remove = FALSE) %>%
      tidyr::separate(col = MLEAF, into = paste0("MLEAF" ,1:x),  sep = ",", remove = FALSE) %>%
      tidyr::separate(col = MLEAC, into = paste0("MLEAC" ,1:x),  sep = ",", remove = FALSE)
    
    tmp.list <- lapply(1:x, FUN = function(i){ 
    
        tmp.split <- dplyr::mutate(tmp,  gt_alt_AD = as.numeric(get(paste0("gt_alt_AD", i))) ,
                                         gt_AR     = as.numeric(get(paste0("gt_alt_AD", i)))/gt_DP, 
                                         ID_old    = ID,
                                         gt_AD_old = gt_AD, gt_AD = paste0(gt_ref_AD, ",", get(paste0("gt_alt_AD", i))),
                                         ALT_old   = ALT,   ALT   = get(paste0("ALT", i)),
                                         AC_old    = AC,    AC    = get(paste0("AC", i)),
                                         AF_old    = AF,    AF    = get(paste0("AF", i)),
                                         MLEAF_old = MLEAF, MLEAF = get(paste0("MLEAF", i)),
                                         MLEAC_old = MLEAC, MLEAC = get(paste0("MLEAC", i)),
                                         ID = paste(CHROM, POS, REF, ALT, sep="_"),
                                   ) %>%
        dplyr::select(-gt_ref_AD, 
                      -paste0("ALT", 1:x), 
                      -paste0("gt_alt_AD", 1:x),  
                      -paste0("AC", 1:x),
                      -paste0("AF", 1:x),
                      -paste0("MLEAF", 1:x),
                      -paste0("MLEAC", 1:x)) %>%
        dplyr::filter(ALT != "*")
        
        })
    
        return(tmp.list %>% purrr::reduce(rbind))
    })
  
    return(purrr::reduce(vcfR.tid.list, rbind))
}
```

```{r , echo=FALSE, warning=FALSE, error=FALSE}
haplotype_caller<-paste0(workspace,"/03_snv_indels_calling/",config_ad$project_name,"_snv_indel.vcf")
vcfR.inst <- vcfR::read.vcfR(haplotype_caller)

vcfR.tid.list <- vcfR::vcfR2tidy(vcfR.inst)

vcfR.tid <- full_join(vcfR.tid.list$fix, vcfR.tid.list$gt, by = c("ChromKey", "POS")) %>% 
            dplyr::mutate(ID = dplyr::if_else(is.na(ID), paste(CHROM, POS, REF, ALT, sep="_"), ID)) 

variants <- vcfR_split_multiAlt(vcfR.tid) %>%
  dplyr::mutate(gt_AR_corr = dplyr::if_else(gt_AR < AR.min, 0, gt_AR),
         gt_AR_corr = dplyr::if_else(gt_DP < 4,   0, gt_AR_corr),
         gt_DP_corr = dplyr::if_else(gt_DP < 4,   0, as.numeric(gt_DP)),
         var_type   = dplyr::if_else(str_length(REF) == str_length(ALT), "SNV", dplyr::if_else(str_length(REF) > str_length(ALT), "DEL", "INS")),
         var_length = str_length(ALT) - str_length(REF),
         var_class  = dplyr::if_else(var_length == 0, var_type, 
                      dplyr::if_else(abs(var_length) == 1,  paste0(var_type, "1"),
                      dplyr::if_else(abs(var_length) >  50, paste0(var_type, ">50"), paste0(var_type, "<50"))))) 

variants.control <- variants %>% filter(Indiv %in% dplyr::filter(sample_sheet, condition == "WT")$plant, gt_AR_corr > 0) 
variants.freq_corr <- variants %>% filter(gt_AR_corr > 0, !is.na(gt_AR_corr)) %>% 
  select(ID, Indiv, target,amplicon) %>% unique() %>% group_by(ID, target,amplicon) %>% count(name="freq")

variants <- variants %>% dplyr::filter(!ID %in% variants.control$ID) %>% dplyr::full_join(variants.freq_corr, by = c("ID", "target","amplicon"))
#DT::datatable(variants)
```
