---
title: "MultiCrispr Project: Data analysis"
author: "Gabri√®le Adam"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
    html_document:
        highlight: tango 
        number_sections: no
        theme: default
        toc: yes
        toc_depth: 3
        toc_float:
            collapsed: no
            smooth_scroll: yes
params:
    result_dir: result_dir
    rmd: "report.Rmd"
    sample: sample
    DP.min: DP.min
    AR.min: AR.min
    guide: guide
    region: region
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load_library, echo=FALSE, warning=FALSE, error=FALSE}
library(tidyverse)
library(DT)
library(vcfR)
library(ggplot2)
```

```{r define_params, echo=FALSE, warning=FALSE, error=FALSE}
DP.min<-params$DP.min
AR.min<-params$AR.min
workspace<-params$result_dir
barcodeDesign<-params$sample
guide<-params$guide
region<-params$region
```

# Introduction

# Data / Input

```{r read_input, echo=FALSE, warning=FALSE, error=FALSE}

barcode_sheet <- read.csv(file = barcodeDesign, header = TRUE, sep = ",") %>% mutate(barcodeName = paste0(barcode_R_name, "--", barcode_F_name)) 
plant.names <- barcode_sheet$plant
controls <- dplyr::filter(barcode_sheet, condition == "WT")$plant
cases    <- dplyr::filter(barcode_sheet, condition != "WT")$plant
```
* number of plants : `r length(unique(plant.names))`
* number of wells : `r length(unique(barcode_sheet$well))`
* number of conditions : `r unique(barcode_sheet$condition)`

# Analysis

## Demultiplexing results

```{r, echo=FALSE, warning=FALSE, error=FALSE}
ccs.lima.log <- read.table(paste0(workspace,"/tmp/lima_summary_extract"), sep="\t")
```
* number of CCS multiplexed reads : `r ccs.lima.log[1,1]`
* number of CCS demultiplexed reads : `r ccs.lima.log[2,1]`

## Mapping

```{r mapping_flagstat, echo=FALSE, warning=FALSE, error=FALSE,  out.width="50%", fig.show="hold"}


  file <- paste0(workspace,"/03_mapping/flagstat/Extract_data.flagstat")
  if(file.exists(file)){
    
    flagstat <- read.table(file, sep="\t", header=FALSE) %>% 
      dplyr::mutate(mapped=V3-V4, unmapped=V2-V3, supplementary=V4, 
                    mapped.ratio=round(mapped/(mapped+unmapped)*100,2), 
                    unmapped.ratio=round(unmapped/(mapped+unmapped)*100,2),
                    supplementary.ratio=round(supplementary/(mapped+unmapped)*100,2)) %>% dplyr::select(-V2, -V3, -V4)
    names(flagstat) <- c("well", "mapped", "unmapped", "supplementary", "mapped.ratio", "unmapped.ratio", "supplementary.ratio")

  } 

DT::datatable(flagstat, 
              options = list(rownames = TRUE, pageLength = 10, scrollX = T, dom = 'tip'),
              class = 'cell-border stripe')
#count
nbr.reads.melt <- flagstat %>% dplyr::select(well, mapped, unmapped, supplementary) %>% reshape2::melt(id = "well", value.name = "nbr.reads")
summary.nbr <- nbr.reads.melt %>% group_by(variable) %>% summarise(median=median(nbr.reads), max=max(nbr.reads), min=min(nbr.reads))
nbr.reads.melt <- left_join(nbr.reads.melt, summary.nbr, by = "variable") %>% mutate(x.label=paste0(variable, "\nmedian=", median, "\nmax=", max, "\nmin=", min))
ggplot(data = nbr.reads.melt, aes(y=nbr.reads, x=x.label)) + geom_boxplot()  + xlab("") + ylab("number of reads") + ggtitle("Boxplot")
#ratio
ratio.reads.melt <- flagstat %>% dplyr::select(well, mapped.ratio, supplementary.ratio) 
names(ratio.reads.melt) <- c("well", "mapped", "supplementary")
ratio.reads.melt <- ratio.reads.melt %>% reshape2::melt(id = "well", value.name = "ratio")
summary.ratio <- ratio.reads.melt %>% group_by(variable) %>% summarise( median=round(median(ratio),2), max=round(max(ratio),2), min=round(min(ratio),2))
ratio.reads.melt <- left_join(ratio.reads.melt, summary.ratio, by = "variable") %>% mutate(x.label=paste0(variable, "\nmedian=", median, "%\nmax=", max, "%\nmin=", min, "%"))
ggplot(data = ratio.reads.melt) + geom_boxplot(aes(y=ratio, x=x.label)) + xlab("") + ylab("% reads") + scale_y_continuous(breaks = seq(0, max(ratio.reads.melt$ratio)+10, 20)) + ggtitle("Boxplot2")
```

```{r coverage_analysis, echo=FALSE, warning=FALSE, error=FALSE, out.width="50%", fig.show="hold"}
bed.tab <- read.table(text=gsub(",","\t",readLines(region)))
names(bed.tab) <- c("rname", "start", "stop", "target","amplicon")
```
### Coverage analysis ontarget regions sequences {.tabset .tabset-fade}

```{r plot_coverage, results='asis',echo=TRUE, warning=FALSE, error=FALSE}

cov.list <- lapply(barcode_sheet$well, function(x){

  file <- paste0(workspace,"/03_mapping/coverage/", x, ".coverage")
  if(file.exists(file)){
    
    read.table(file, sep="\t") %>%  dplyr::filter(V4 > 1)  %>% dplyr::mutate(well = x)
  }
})

cov <- purrr::reduce(cov.list, rbind) %>% dplyr::mutate(nbr_reads=dplyr::if_else(V4 >= 15, ">15X", "<15X"))
colnames(cov) <- c("rname",	"startpos",	"endpos",	"numreads",	"covbases",	"coverage",	"meandepth",	"meanbaseq",	"meanmapq", "well", "nbr_reads")
cov <- cov %>% left_join(bed.tab, by=c("rname","startpos"="start","endpos"="stop")) 
cov.summary <- cov %>% group_by(amplicon,target) %>% summarise(min=min(numreads), max=max(numreads), median=round(median(numreads)))
cov <- left_join(cov, cov.summary, by = c("amplicon","target")) %>% mutate(x.label=paste0(target, "\nmedian=", median, "\nmax=", max, "\nmin=", min)) 
for (i in unique(bed.tab$target)){
cat("####",i,"\n")
g<-ggplot(filter(cov,target==i)) + geom_boxplot(aes(y=numreads, x=x.label)) + xlab("") + ylab("#reads") + ggtitle("Plot coverage")+facet_wrap(. ~ target,scales = "free")+theme(axis.text.x=element_text(angle=90,hjust=1))
print(g)
cat('\n\n')
}
```
</div>

```{r plot_depth}
cov.list <- lapply(barcode_sheet$well, function(x){

  file <- paste0(workspace, "/03_mapping/depth/", x, ".depth")
  if(file.exists(file)){

    read.table(file, sep="\t")  %>% dplyr::mutate(plant = x)
  }
})

depth <- purrr::reduce(cov.list, rbind)
colnames(depth) <- c("rname", "position", "depth", "plant")
depth <- depth %>% left_join(bed.tab, by="rname")  %>% filter(position >=start, position <= stop)
heading<-as.character(unique(bed.tab$target))
```

### Representation for each target region by amplicon {.tabset .tabset-fade}

```{r create_fct_for_tabulation, results='asis',echo =TRUE}

for (i in unique(bed.tab$target)){
 cat("####",i,"\n")
  p <- ggplot(data = filter(depth,target== i)) + geom_line(aes(x=position, y=depth, color=plant))   + theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylab("Depth of coverage")+facet_grid(~amplicon, scales = "free") 
  print(p)
cat('\n\n')
}
```

</div>

### Proportion of {.tabset .tabset-fade}

```{r proportion_of, results='asis',echo=TRUE, warning=FALSE, error=FALSE}
for (i in unique(bed.tab$target)){
 cat("####",i,"\n")
tmp <- cov  %>% filter(target== i) %>% group_by(nbr_reads, target, amplicon) %>% count(name = "#well") 
g1<-ggplot(data = tmp) + geom_bar(aes(x=amplicon, y=`#well`, fill=nbr_reads), stat = "identity") + xlab("") +facet_grid(~target, scales = "free")
g2<-cov %>% filter(target== i) %>% group_by(well, nbr_reads,target) %>% count(name = "amplicon") %>% filter(nbr_reads == ">15X") %>% group_by(amplicon,target) %>% count(name = "#well") %>% ggplot(aes(x = amplicon, y = `#well`)) + geom_bar(stat = "identity") + xlab("nbr of amplicon covered at >15X")
print(g1)
print(g2)
cat('\n\n')
}
```
</div>

## SNV and Indels calling

```{r define_r_fct_multiallele, echo=FALSE}
vcfR_split_multiAlt <- function(vcfR.tid, nbrAlt=NULL){

  vcfR.tid.tmp <- 
    dplyr::mutate(vcfR.tid, nb.allele=unlist(lapply(str_split(string = vcfR.tid$gt_AD, pattern = ","), function(x){return(length(x))}))) %>%
    dplyr::mutate(nb.allele=nb.allele-1)
  
  vcfR.tid.list <- lapply(unique(vcfR.tid.tmp$nb.allele), function(x){
    
    tmp <- vcfR.tid.tmp %>% dplyr::filter(nb.allele == x) %>%
      tidyr::separate(col = gt_AD, into = c("gt_ref_AD", paste0("gt_alt_AD",1:x)), sep = ",", remove = FALSE) %>%
      tidyr::separate(col = ALT,   into = paste0("ALT" ,1:x), sep = ",", remove = FALSE) %>%
      tidyr::separate(col = AC,    into = paste0("AC" ,1:x),  sep = ",", remove = FALSE) %>%
      tidyr::separate(col = AF,    into = paste0("AF" ,1:x),  sep = ",", remove = FALSE) %>%
      tidyr::separate(col = MLEAF, into = paste0("MLEAF" ,1:x),  sep = ",", remove = FALSE) %>%
      tidyr::separate(col = MLEAC, into = paste0("MLEAC" ,1:x),  sep = ",", remove = FALSE)
    
    tmp.list <- lapply(1:x, FUN = function(i){ 
    
        tmp.split <- dplyr::mutate(tmp,  gt_alt_AD = as.numeric(get(paste0("gt_alt_AD", i))) ,
                                         gt_AR     = as.numeric(get(paste0("gt_alt_AD", i)))/gt_DP, 
                                         ID_old    = ID,
                                         gt_AD_old = gt_AD, gt_AD = paste0(gt_ref_AD, ",", get(paste0("gt_alt_AD", i))),
                                         ALT_old   = ALT,   ALT   = get(paste0("ALT", i)),
                                         AC_old    = AC,    AC    = get(paste0("AC", i)),
                                         AF_old    = AF,    AF    = get(paste0("AF", i)),
                                         MLEAF_old = MLEAF, MLEAF = get(paste0("MLEAF", i)),
                                         MLEAC_old = MLEAC, MLEAC = get(paste0("MLEAC", i)),
                                         ID = paste(CHROM, POS, REF, ALT, sep="_"),
                                   ) %>%
        dplyr::select(-gt_ref_AD, 
                      -paste0("ALT", 1:x), 
                      -paste0("gt_alt_AD", 1:x),  
                      -paste0("AC", 1:x),
                      -paste0("AF", 1:x),
                      -paste0("MLEAF", 1:x),
                      -paste0("MLEAC", 1:x)) %>%
        dplyr::filter(ALT != "*")
        
        })
    
        return(tmp.list %>% purrr::reduce(rbind))
    })
  
    return(purrr::reduce(vcfR.tid.list, rbind))
}
```

```{r , echo=FALSE, warning=FALSE, error=FALSE}

vcf_file<-paste0(workspace,"/04_snv_indels/genotype_plant.annotate.vcf")

vcfR.inst <- vcfR::read.vcfR(vcf_file)

vcfR.tid.list <- vcfR::vcfR2tidy(vcfR.inst)

vcfR.tid <- full_join(vcfR.tid.list$fix, vcfR.tid.list$gt, by = c("ChromKey", "POS")) %>% 
            dplyr::mutate(ID = dplyr::if_else(is.na(ID), paste(CHROM, POS, REF, ALT, sep="_"), ID)) 


variants <- vcfR_split_multiAlt(vcfR.tid) %>% left_join(bed.tab, by=c("CHROM" = "rname")) %>% filter(POS >=start, POS <= stop) %>% select(-stop, -start) %>%
  dplyr::mutate(gt_AR_corr = dplyr::if_else(gt_AR < AR.min, 0, gt_AR),
         gt_AR_corr = dplyr::if_else(gt_DP < 4,   0, gt_AR_corr),
         gt_DP_corr = dplyr::if_else(gt_DP < 4,   0, as.numeric(gt_DP)),
         var_type   = dplyr::if_else(str_length(REF) == str_length(ALT), "SNV", dplyr::if_else(str_length(REF) > str_length(ALT), "DEL", "INS")),
         var_length = str_length(ALT) - str_length(REF),
         var_class  = dplyr::if_else(var_length == 0, var_type, 
                      dplyr::if_else(abs(var_length) == 1,  paste0(var_type, "1"),
                      dplyr::if_else(abs(var_length) >  50, paste0(var_type, ">50"), paste0(var_type, "<50"))))) 

variants.control <- variants %>% filter(Indiv %in% dplyr::filter(barcode_sheet, condition == "WT")$plant, gt_AR_corr > 0) 
variants.freq_corr <- variants %>% filter(gt_AR_corr > 0, !is.na(gt_AR_corr)) %>% 
  select(ID, Indiv, target,amplicon) %>% unique() %>% group_by(ID, target,amplicon) %>% count(name="freq")

variants <- variants %>% dplyr::filter(!ID %in% variants.control$ID) %>% dplyr::full_join(variants.freq_corr, by = c("ID", "target","amplicon"))
#DT::datatable(variants)
```

### nbr of variants per target and individu {.tabset .tabset-fade}

```{r plot_nbr_variant ,results='asis',echo=TRUE}
for (i in unique(bed.tab$target)){
  cat("####",i,"\n")
g<-variants  %>% filter(gt_AR_corr > 0, !is.na(gt_AR_corr), target==i)  %>% 
  select(ID, Indiv, target,amplicon) %>% unique() %>% group_by(Indiv, target,amplicon) %>% count(name="nbr_variants") %>%
  ggplot() + geom_boxplot(aes(y=nbr_variants, x=amplicon))
print(g)
g2<- variants  %>% filter(gt_AR_corr > 0, !is.na(gt_AR_corr), target==i)  %>%
  select(ID, Indiv, target, var_type, amplicon) %>% unique() %>% group_by(Indiv, target, var_type,amplicon) %>% count(name="nbr_variants") %>%
  ggplot() + geom_boxplot(aes(y=nbr_variants, x=var_type)) + facet_grid(~amplicon)
print(g2)
g3<- variants  %>% filter(gt_AR_corr > 0, !is.na(gt_AR_corr),target==i)  %>%
  select(ID, Indiv, target, var_class,amplicon) %>% unique() %>% group_by(Indiv, target, var_class,amplicon) %>% count(name="nbr_variants") %>%
  ggplot() + geom_boxplot(aes(y=nbr_variants, x=target)) + facet_grid(amplicon~var_class) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylim(0, 10)
print(g3)
cat('\n\n')
}

```
</div>

### count per variant type and Indel length -Allele frequency in genotypes {.tabset .tabset-fade}

```{r count_allele_by_type,results='asis',echo=TRUE}
for (i in unique(bed.tab$target)){
cat("####",i,"\n")
g1<-variants %>% filter(gt_AR_corr > 0, !is.na(gt_AR_corr),target==i) %>%  select(ID, Indiv, target) %>% unique() %>% group_by(ID, target) %>% count(name="freq") %>% ggplot() + 
	geom_bar(aes(as.factor(freq), fill=target)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
print(g1)
g2<-variants %>% filter(gt_AR_corr > 0, !is.na(gt_AR_corr),target==i) %>% select(ID, Indiv, var_type) %>% unique() %>% group_by(ID, var_type) %>% count(name="freq") %>% ggplot() + geom_bar(aes(as.factor(freq), fill=var_type)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
print(g2)
g3<-variants %>% filter(gt_AR_corr > 0, !is.na(gt_AR_corr),target==i) %>% select(ID, Indiv, var_class) %>% unique() %>% group_by(ID, var_class) %>% count(name="freq") %>% ggplot() + geom_bar(aes(as.factor(freq), fill=var_class)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
print(g3)
g4<-variants %>% filter(gt_AR_corr > 0, !is.na(gt_AR_corr),target==i) %>% select(ID, Indiv, var_class) %>% unique() %>% group_by(ID, var_class) %>% count(name="freq") %>% ggplot() + geom_bar(aes(as.factor(freq), fill=var_class)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
cat('\n\n')
}
```
</div>

## SV Calling

```{r study_sv_calling }
vcf_file<-paste0(workspace,"/05_sv_calling/variants.pbsv.annotate.vcf")
sv.vcf <- vcfR::read.vcfR(vcf_file)

vcfR.tid.sv.list <- vcfR::vcfR2tidy(sv.vcf)

vcfR.tid.sv <- dplyr::full_join(vcfR.tid.sv.list$fix, vcfR.tid.sv.list$gt, by = c("ChromKey", "POS")) %>% 
            dplyr::mutate(ID = dplyr::if_else(is.na(ID), paste(CHROM, POS, REF, ALT, sep="_"), ID))  %>%
            left_join(bed.tab, by=c("CHROM" = "rname"))%>% filter(POS >=start, POS <= stop) %>% select(-stop, -start) %>%
            tidyr::separate(col = gt_AD, into = c("gt_ref_AD", "gt_alt_AD"), sep = ",", remove = FALSE) %>%
            dplyr::mutate(gt_AR      = as.numeric(gt_alt_AD)/gt_DP,
                          gt_AR_corr = dplyr::if_else(gt_AR < 0.2, 0, gt_AR),
                          gt_AR_corr = dplyr::if_else(gt_DP < 4,   0, gt_AR_corr),
                          gt_DP_corr = dplyr::if_else(gt_DP < 4,   0, as.numeric(gt_DP)))
                                          
DT::datatable(vcfR.tid.sv)
```

### Number of variant per target {.tabset .tabset-fade}
```{r plot_sv_calling ,results='asis',echo=TRUE} 
for (i in unique(bed.tab$target)){
cat("####",i,"\n")
g1<-vcfR.tid.sv  %>% filter(gt_AR_corr > 0, !is.na(gt_AR_corr),target==i)  %>% 
  select(ID, Indiv, target,amplicon) %>% unique() %>% group_by(Indiv, target,amplicon) %>% count(name="nbr_variants") %>%
  ggplot() + geom_boxplot(aes(y=nbr_variants, x=amplicon))+facet_grid(~target, scales = "free")
print(g1)
g2<-vcfR.tid.sv  %>% filter(gt_AR_corr > 0, !is.na(gt_AR_corr),target==i)  %>% 
  select(ID, target,amplicon, SVLEN) %>% unique() %>% group_by(target,amplicon, SVLEN) %>% count(name="sv_length") %>%
  ggplot() + geom_boxplot(aes(y=sv_length, x=amplicon))+facet_grid(~target, scales = "free")
#print(g2)
g3<-vcfR.tid.sv  %>% filter(gt_AR_corr > 0, !is.na(gt_AR_corr),target==i)  %>%
  select(ID, target, SVLEN,amplicon) %>% unique() %>% group_by(target, SVLEN,amplicon) %>%
  ggplot() + geom_point(aes(y=SVLEN, x=amplicon))+facet_grid(~target, scales = "free")
#print(g3)
cat('\n\n')
}
```
</div>

# Source
<a download="Report_snpeff_pbsv.html" href="`r base64enc::dataURI(file =params$rmd, mime = 'text/rmd', encoding = 'base64')`">HTML report obtained after snpeff on pbsv data</a>
<a download="report.Rmd" href="`r base64enc::dataURI(file = params$rmd, mime = 'text/rmd', encoding = 'base64')`">R Markdown source file (to produce this document)</a>
<a download="Report_snpeff.html" href="`r base64enc::dataURI(file = params$rmd, mime = 'text/rmd', encoding = 'base64')`">HTML report obtained after snpeff on GenomicsVCF data</a>

