# SNV and small Indels



```{r}

vcfR <- vcfR::read.vcfR(vcf.file, verbose = FALSE)

vcfR.tb.list <- vcfR::vcfR2tidy(vcfR, verbose = FALSE)

vcfR.tb <- full_join(vcfR.tb.list$fix, vcfR.tb.list$gt, by = c("ChromKey", "POS")) %>% 
  dplyr::mutate(ID = dplyr::if_else(is.na(ID), paste(CHROM, POS, REF, ALT, sep="_"), ID)) 

# split multiallelic variants + add columns  with var type information
variants.gatk <- vcfR_split_multiAlt(vcfR.tb) %>% 
  dplyr::mutate(var_length = str_length(ALT) - str_length(REF),
                var_type   = dplyr::if_else(var_length == 0, "SNV", 
                             dplyr::if_else(var_length  < 0, "DEL", "INS")),
                ID = paste(CHROM, POS, var_type, var_length, sep = "_"),
                tools = "GATK")

#!!!!!! CA    AA   
```

## Quality control and filtering

> Before Filtering

```{r score quality, out.width="50%", fig.show="hold", message=FALSE, warning=FALSE, echo=FALSE}

# Score de qualité
p <- variants.gatk %>% select(ID, QUAL) %>% unique() %>% ggplot() + geom_density(aes(QUAL)) + xlim(0,200) +
  xlab("quality score") + geom_vline(xintercept = qual.min, color = "red") + ggtitle("Variant quality score distribution")
print(p)

vcf.tbl.g   <- dplyr::filter(variants.gatk,   gt_AR != 0, !is.na(gt_AR))

# distribution de la couverture des positions variantes
p <- ggplot(data=vcf.tbl.g) + geom_density(aes(x=gt_DP, color=Indiv)) + xlim(0,50) +
  ylab("DP of variants") + geom_vline(xintercept = DP.min, color = "red") + ggtitle("Variant depth of coverage distribution")
if (length(unique(vcf.tbl.g$Indiv)) > 15){
  p <- p + theme(legend.position="none")
}

print(p)

# Distribution de l'allèle ratio
p <- ggplot(data=vcf.tbl.g) + geom_density(aes(gt_AR, color=Indiv)) + 
  xlab("Alternative allele ratio") + xlim(0,1) + geom_vline(xintercept = AR.min, color = "red")+ ggtitle("Variant allele ratio distribution")

if (length(unique(vcf.tbl.g$Indiv)) > 15){
  p <- p + theme(legend.position="none")
}

print(p)

# compter le nombre de positions non couvertes par échantillon.
p <- ggplot(data = vcf.tbl.g) +  geom_bar(aes(x=Indiv)) + 
  theme(axis.text.x = element_text(angle = 90, size = rel(0.8), hjust = 0)) + xlab("") + 
  ylab("Non covered positions")+ ggtitle("Non covered variant positions")
print(p)

```

```{r var depth, out.width="50%", fig.show="hold"}
NA_ID <- filter(variants.gatk, nb_NA > 0)$ID

variants.gatk %>% filter(ID %in% NA_ID) %>% 
  ggplot() +  geom_bar(aes(x=as.factor(nb_NA))) + 
  theme(axis.text.x = element_text(angle = 90, size = rel(0.8), hjust = 0)) + xlab("") + 
  ylab("Non covered positions") + ggtitle("Non covered variant positions")


```


> filtering threshold

* score de qualité : > `r qual.min`
* nombre de reads qui couvrent la position variant : > `r DP.min`
* nombre de reads qui portent l'allèle alternatif : > 1
* min AR : `r AR.min`
* retirer les positions avec au moins un échantillon non couvert.



```{r qc filtre}

# filter variants with low quality/coverage + add amplicon positions
variants.gatk <- filter(variants.gatk, nb_NA == 0)
variants.gatk.flt <- qc_filtering(variants.gatk, DP.min, AR.min, qual.min)

# dplyr::select(CHROM, POS, ID, REF, ALT, QUAL, EFF, Indiv, gt_GT, gt_AD, gt_DP, gt_AR, 
#               var_length, var_type,  amplicon, locus, strand, freq, tools) 

```



## Descriptive statistics

### Number of small variants

```{r nb_SNV, out.width="90%"}

vcf.tbl.q.g <- filter(variants.gatk.flt, gt_AR != 0, !is.na(gt_AR))
vcf.tbl.q.g %>% dplyr::filter(freq != length(unique(sample_sheet$sample))) %>% dplyr::select(ID, gt_AR, Indiv, var_type) %>% unique() %>% 
  ggplot() + geom_bar(aes(x=Indiv, fill=var_type)) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
  xlab("") + ylab("nb of variants")
```

### Variant frequency

```{r SNV_freq, out.width="90%"}

p <- vcf.tbl.q.g %>% dplyr::select(Indiv, freq, ID) %>% group_by(Indiv, freq) %>% 
  dplyr::filter(freq != length(unique(sample_sheet$sample))) %>% dplyr::count(name = "nb_var") %>% 
  ggplot(aes(x=as.factor(freq), y=nb_var)) + geom_boxplot(outlier.shape = NA) +
  geom_point(size = 0.8, aes(color=Indiv), position=position_jitter(width = 0.2)) + xlab("freq in samples") + ylab("# variants")

if (length(unique(vcf.tbl.q.g$Indiv)) > 15){
  
  p <- p + theme(legend.position="none")
}
print(p)

vcf.tbl.g %>% filter(POS == 1548127)

# warning -> samples with the same variants or pas de variant
```

### Clustering

```{r SNV_clustering, out.width="90%"}

df <- vcf.tbl.q.g %>% dplyr::filter( freq != length(unique(sample_sheet$sample))) %>% dplyr::select(ID, gt_AR, Indiv) %>% 
  reshape2::dcast(ID~Indiv , value.var = "gt_AR") 

# calcul la distance entre les lignes de la matrice
# vcf.tidy.bin[,-1] : sans la 1ere colonne qui contient les IDs
dist  <- dist(t(df[,-1]), method = "euclidean")

# clustering hiérarchique
hc    <- hclust(dist)

# arranger la structure du dendrogramme
dend  <- as.dendrogram(hc)

# dessiner le dendrogramme
plot(dend, main = "Sample clustering using all variants")
```


### 3 type de substitution for AG > TA ...

```{r}

p <- vcf.tbl.q.g %>% dplyr::filter(var_type == "SNV", str_length(ALT) == 1) %>% 
  dplyr::select(ALT, REF, Indiv) %>% group_by(ALT, REF, Indiv) %>% 
  dplyr::count(name = "nb_sub") %>% 
  ggplot(aes(x=paste(REF, ALT, sep=">"), y=nb_sub)) + geom_boxplot(outlier.shape = NA) +
  geom_point(size = 0.8, aes(color=Indiv), position=position_jitter(width = 0.2))

if (length(unique(vcf.tbl.q.g$Indiv)) > 15){
  
  p <- p + theme(legend.position="none")
}
print(p)

```




### 4 annotations
```{r}

colname.ANN <- c("Allele", "effect", "impact", "GeneName", 
                 "GeneID", "FeatureType", "FeatureID", "TranscriptBiotype", 
                 "Rank", "HGVS.c", "HGVS.p", "cDNA_len:", 
                 "CDS_len", "Protein_len", "DistanceToFeature", "Warnings")

exlude_effect <- c("upstream_gene_variant", "downstream_gene_variant", "intergenic_region", "intragenic_variant", "intragenic_variant", 
                   "3_prime_UTR_variant", "", "", "")

ID.ANN <- vcf.tbl.q.g %>% dplyr::select(ID, ANN) %>% unique() 


ANN.split <- lapply(1:dim(ID.ANN)[1], FUN = function(x){
  
  tmp <- ID.ANN$ANN[x] %>% str_split(",") %>% unlist()
  data.frame(ID = rep(ID.ANN$ID[x], length(tmp)), ANN = tmp)
  
}) %>% purrr::reduce(rbind) %>% 
  tidyr::separate(col = ANN, into = colname.ANN, sep = "[|]") %>% 
  dplyr::filter(!effect %in% exlude_effect) %>% 
  dplyr::select(ID, effect) %>% unique()

effect.split <- lapply(1:dim(ANN.split)[1], function(x){
  
  tmp <- ANN.split$effect[x] %>% stringr::str_split("&") %>% unlist()
  data.frame(ID = rep(ANN.split$ID[x], length(tmp)), effect = tmp)
}) %>% purrr::reduce(rbind)


effect.split %>% dplyr::group_by(effect) %>% dplyr::count()

```

